### PR TITLE
[core][autoscaler] ray status - ray cluster status formatter [3/x]

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR adds the formatting logic (that still uses legacy `format_info_string` to be compatible) with `ray status`. 

In future refactoring, the conversion implementation will be changed to format the info directly as we migrate away from v1. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/utils.py [hunk 1]
```diff
@@ -28,6 +36,163 @@
...
+        )
+
+    @classmethod
+    def _parse_autoscaler_summary(cls, data: ClusterStatus) -> AutoscalerSummary:
+        active_nodes = _count_by(data.healthy_nodes, "ray_node_type_name")
+        pending_launches = _count_by(data.pending_launches, "ray_node_type_name")
+        pending_nodes = []
+        for node in data.pending_nodes:
+            # We are using details for the pending node's status.
+            # TODO(rickyx): we should probably use instance id rather than ip address
```
Comment (rkooo567) [inline python/ray/autoscaler/v2/utils.py:88]: maybe both? is there a way to mapping instance id -> IP address?

File: python/ray/autoscaler/v2/utils.py [hunk 2]
```diff
@@ -28,6 +36,163 @@
...
+
+        # From IP to node type name.
+        node_type_mapping = {}
+        for node in data.healthy_nodes:
+            node_type_mapping[node.ip_address] = node.ray_node_type_name
+
+        # Transform failed launches to node_availability_summary
+        node_availabilities = {}
+        for failed_launch in data.failed_launches:
+            # TODO(rickyx): we could also add failed timestamp, count info.
```
Comment (rkooo567) [inline python/ray/autoscaler/v2/utils.py:108]: is it existing in the V1 status?

File: python/ray/autoscaler/v2/utils.py [hunk 3]
```diff
@@ -28,6 +36,163 @@
...
+        for demand in data.resource_demands.ray_task_actor_demand:
+            for bundle_by_count in demand.bundles_by_count:
+                resource_demands.append((bundle_by_count.bundle, bundle_by_count.count))
+
+        pg_demand = []
+        pg_demand_strs = []
+        pg_demand_str_to_demand = {}
+        for pg_demand in data.resource_demands.placement_group_demand:
+            s = pg_demand.strategy + "|" + pg_demand.state
+            pg_demand_strs += [s]
```
Comment (rkooo567) [inline python/ray/autoscaler/v2/utils.py:149]:
```suggestion
            pg_demand_strs.append(s)
```

File: python/ray/autoscaler/v2/utils.py [hunk 4]
```diff
@@ -28,6 +36,163 @@
...
+            # TODO(rickyx): we should probably use instance id rather than ip address
+            # here.
+            pending_nodes.append(
+                (node.ip_address, node.ray_node_type_name, node.details)
+            )
+
+        failed_nodes = []
+        for node in data.failed_nodes:
+            # TODO(rickyx): we should probably use instance id/node id rather
+            # than node type here since node type is not unique among failed nodes.
```
Comment (jjyao) [inline python/ray/autoscaler/v2/utils.py:97]: You mean node ip not node type?

File: python/ray/autoscaler/v2/utils.py [hunk 5]
```diff
@@ -194,7 +374,10 @@ def _parse_cluster_resource_usage(
         cluster_resource_usage = defaultdict(ResourceUsage)
 
         for node_state in state.node_states:
-            cls._parse_node_resource_usage(node_state, cluster_resource_usage)
+            if node_state.status != NodeStatus.DEAD:
```
Comment (rkooo567) [inline python/ray/autoscaler/v2/utils.py:377]: what's the reason behind this change? I think it is probably better still having cluster resources of the dead node?

**Review by rkooo567 — APPROVED** [review]
LGTM. Only nit comments;

File: python/ray/autoscaler/v2/tests/test_utils.py [hunk 6]
```diff
@@ -231,6 +303,221 @@ def test_cluster_status_parser_autoscaler_state():
...
+            none_terminated_node_request_time_s=0.2,
+            autoscaler_iteration_time_s=0.3,
+            autoscaler_version="10",
+            cluster_resource_state_version="20",
+            request_timestamp_s=775303535,
+        ),
+    )
+    actual = ClusterStatusFormatter.format(state, verbose=True)
+
+    expected = """======== Autoscaler status: 1994-07-27 10:05:35 ========
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply python/ray/autoscaler/v2/tests/test_utils.py:472]: the existing tests are all mocked unit tests like this, so it's not that easy to test the parity directly. 

But I plan to add e2e tests for ray status printout in a follow-up PR, and I can test it in there.

File: python/ray/autoscaler/v2/utils.py (hunk 2 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/utils.py:108]: nope.

File: python/ray/autoscaler/v2/utils.py (hunk 5 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/utils.py:377]: hmm, but a dead node technically doesn't contribute to any cluster resources? (e.g. it will be confusing to see 10 CPUs total while all 10s are from a dead node? As the underlying instance might also be gone)

File: python/ray/autoscaler/v2/utils.py (hunk 4 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/utils.py:97]: yeah, right.

File: python/ray/autoscaler/v2/utils.py (hunk 5 above)
Comment (rkooo567) [reply python/ray/autoscaler/v2/utils.py:377]: Are we showing this from the ray status? (individual dead node)?

File: python/ray/autoscaler/v2/tests/test_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/autoscaler/v2/tests/test_utils.py:472]: maybe we can just provide the same input from existing unit tests? I just want to make sure we don't introduce regression by not testing the scenario tested from existing unit tests.

File: python/ray/autoscaler/v2/utils.py (hunk 5 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/utils.py:377]: yes - not resources usage. but node (ip, node_id, type) as failed nodes.

File: python/ray/autoscaler/v2/tests/test_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rickyyx) [reply python/ray/autoscaler/v2/tests/test_utils.py:472]: the current code actually does call what the existing `ray status` called. So prepping the same input would essentially not trigger what's been changed in v2. 

but I have a pr that tests v1/v2 parity explicitly.

[the remark this answers is not in the document]
Comment (rickyyx) [reply python/ray/autoscaler/v2/tests/test_utils.py:472]: https://github.com/ray-project/ray/pull/37473 here.
